File: project/modules/responsiveWebSocketConnection/src/main/java/rws/common/responsiveWebSocketConnection/impl/modules/timeouts/Timeouts.java
package rws.common.responsiveWebSocketConnection.impl.modules.timeouts;

import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import rws.common.responsiveWebSocketConnection.impl.modules.timeouts.Timeout;

public final class Timeouts {
  public static Timeout setTimeout(Runnable fn, int milliseconds) {
    final ScheduledFuture task = _executor.schedule(fn, milliseconds, TimeUnit.MILLISECONDS);
    return new Timeout(task);
  }

  public static void clearTimeout(Timeout timeout) {
    final boolean mayInterruptIfRunning = false;
    timeout._task.cancel(mayInterruptIfRunning);
  }

  private static final ScheduledThreadPoolExecutor _executor = new ScheduledThreadPoolExecutor(1);
}

File: project/modules/tests/src/test/scala/rws/tests/testResponsiveWebSockets/checks/utils/timeouts.scala
package rws.tests.testResponsiveWebSockets.checks.utils;

import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

final object timeouts {
  type Timeout = ScheduledFuture[_];

  def setTimeout(fn: Runnable, milliseconds: Int) = {
    _executor.schedule(fn, milliseconds, TimeUnit.MILLISECONDS);
  }

  def clearTimeout(timeout: Timeout): Unit = { timeout.cancel(false); }

  private val _executor: ScheduledThreadPoolExecutor = new ScheduledThreadPoolExecutor(1);
  _executor.setRemoveOnCancelPolicy(true);
}
